File: frameworks/AgentsMesh/packages/service-interface/src/view-models/runner.ts
// Runner ViewModels — UI-side projection of `proto.runner_api.v1` types. Owned
// in this zero-dep contract layer so web (fromProtoRunner) and desktop
// (electron-adapter projection) share one definition. Snake_case kept because
// host_info ships as a JSON string the adapter parses and many consumers
// (RunnerCard, RunnerLogsCard, useRunnerDetail) read these directly.
export interface RunnerData {
  id: number;
  node_id: string;
  description?: string;
  status: "online" | "offline" | "maintenance" | "busy";
  last_heartbeat?: string;
  current_pods: number;
  max_concurrent_pods: number;
  runner_version?: string;
  is_enabled: boolean;
  visibility?: "organization" | "private";
  registered_by_user_id?: number;
  host_info?: {
    os?: string;
    arch?: string;
    memory?: number;
    cpu_cores?: number;
    hostname?: string;
  };
  available_agents?: string[];
  agent_versions?: Array<{ slug: string; version: string; path?: string }>;
  tags?: string[];
  created_at?: string;
  updated_at?: string;
  active_pods?: Array<{
    pod_key: string;
    status: string;
    agent_status: string;
  }>;
}

export interface RelayConnectionInfo {
  pod_key: string;
  relay_url: string;
  connected: boolean;
  connected_at: string;
}

export interface GRPCRegistrationToken {
  id: number;
  organization_id?: number;
  name?: string;
  labels?: string[];
  single_use?: boolean;
  max_uses?: number;
  used_count?: number;
  expires_at?: string;
  created_by?: number;
  created_at?: string;
}

export interface RunnerListResponse {
  runners: RunnerData[];
  latest_runner_version?: string;
}

export interface RunnerDetailResponse {
  runner: RunnerData;
  relay_connections?: RelayConnectionInfo[];
  latest_runner_version?: string;
}

export interface RunnerLogData {
  id: number;
  request_id: string;
  status: string;
  storage_key?: string;
  size_bytes: number;
  error_message?: string;
  requested_by_id: number;
  download_url?: string;
  created_at?: string;
  completed_at?: string;
}

export interface RunnerPodData {
  id: number;
  pod_key: string;
  organization_id: number;
  runner_id: number;
  agent_slug?: string;
  custom_agent_slug?: string;
  repository_id?: number;
  ticket_id?: number;
  ticket?: {
    id: number;
    slug: string;
    title: string;
  };
  status: string;
  agent_status: string;
  claude_status?: string;
  branch_name?: string;
  sandbox_path?: string;
  session_id?: string;
  source_pod_key?: string;
  resumed_by_pod_key?: string;
  prompt?: string;
  created_by_id?: number;
  created_at?: string;
  updated_at?: string;
  terminated_at?: string;
}

export interface SandboxStatus {
  pod_key: string;
  exists: boolean;
  can_resume: boolean;
  sandbox_path?: string;
  repository_url?: string;
  branch_name?: string;
  current_commit?: string;
  size_bytes?: number;
  last_modified?: number;
  has_uncommitted_changes?: boolean;
  error?: string;
}

export interface RunnerAuthStatus {
  status: "pending" | "authorized" | "expired";
  node_id?: string;
  expires_at?: string;
}

export interface RunnerAuthorizeResponse {
  runner_id: number;
  node_id: string;
  message: string;
}

File: frameworks/AgentsMesh/backend/internal/domain/runner/runner.go
package runner

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"
)

type HostInfo map[string]interface{}

func (hi *HostInfo) Scan(value interface{}) error {
	if value == nil {
		*hi = nil
		return nil
	}
	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return errors.New("unsupported type for HostInfo Scan")
	}
	return json.Unmarshal(bytes, hi)
}

func (hi HostInfo) Value() (driver.Value, error) {
	if hi == nil {
		return nil, nil
	}
	return json.Marshal(hi)
}

type StringSlice []string

func (s *StringSlice) Scan(value interface{}) error {
	if value == nil {
		*s = nil
		return nil
	}
	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return errors.New("unsupported type for StringSlice Scan")
	}
	return json.Unmarshal(bytes, s)
}

func (s StringSlice) Value() (driver.Value, error) {
	if s == nil {
		return nil, nil
	}
	return json.Marshal(s)
}

type AgentVersion struct {
	Slug    string `json:"slug"`
	Version string `json:"version"`
	Path    string `json:"path,omitempty"`
}

type AgentVersionSlice []AgentVersion

func (s *AgentVersionSlice) Scan(value interface{}) error {
	if value == nil {
		*s = nil
		return nil
	}
	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return errors.New("unsupported type for AgentVersionSlice Scan")
	}
	return json.Unmarshal(bytes, s)
}

func (s AgentVersionSlice) Value() (driver.Value, error) {
	if s == nil {
		return nil, nil
	}
	return json.Marshal(s)
}

func (s AgentVersionSlice) GetAgentVersion(slug string) *AgentVersion {
	for _, v := range s {
		if v.Slug == slug {
			return &v
		}
	}
	return nil
}

const (
	RunnerStatusOnline  = "online"
	RunnerStatusOffline = "offline"
	RunnerStatusBusy    = "busy"
)

const (
	VisibilityOrganization = "organization"
	VisibilityPrivate      = "private"
)

type Runner struct {
	ID             int64  `gorm:"primaryKey" json:"id"`
	OrganizationID int64  `gorm:"not null;index" json:"organization_id"`
	NodeID         string `gorm:"size:100;not null" json:"node_id"`
	Description    string `gorm:"type:text" json:"description,omitempty"`

	Status            string     `gorm:"size:50;not null;default:'offline';index" json:"status"`
	LastHeartbeat     *time.Time `json:"last_heartbeat,omitempty"`
	CurrentPods       int        `gorm:"not null;default:0" json:"current_pods"`
	MaxConcurrentPods int        `gorm:"not null;default:5" json:"max_concurrent_pods"`
	RunnerVersion     *string    `gorm:"size:50" json:"runner_version,omitempty"`
	IsEnabled         bool       `gorm:"not null;default:true" json:"is_enabled"`

	AvailableAgents StringSlice `gorm:"type:jsonb" json:"available_agents,omitempty"`

	AgentVersions AgentVersionSlice `gorm:"type:jsonb" json:"agent_versions,omitempty"`

	HostInfo HostInfo `gorm:"type:jsonb" json:"host_info,omitempty"`

	Tags StringSlice `gorm:"type:jsonb;default:'[]'" json:"tags,omitempty"`

	Visibility         string `gorm:"size:20;not null;default:'organization'" json:"visibility"`
	RegisteredByUserID *int64 `json:"registered_by_user_id,omitempty"`

	// mTLS certificate fields (added for gRPC migration)
	CertSerialNumber *string    `gorm:"size:64" json:"cert_serial_number,omitempty"`
	CertExpiresAt    *time.Time `json:"cert_expires_at,omitempty"`

	CreatedAt time.Time `gorm:"not null;default:now()" json:"created_at"`
	UpdatedAt time.Time `gorm:"not null;default:now()" json:"updated_at"`
}

func (Runner) TableName() string {
	return "runners"
}

func (r *Runner) IsOnline() bool {
	return r.Status == RunnerStatusOnline
}

func (r *Runner) CanAcceptPod() bool {
	return r.IsEnabled && r.IsOnline() && r.CurrentPods < r.MaxConcurrentPods
}

func (r *Runner) SupportsAgent(agentSlug string) bool {
	for _, slug := range r.AvailableAgents {
		if slug == agentSlug {
			return true
		}
	}
	return false
}

func (r *Runner) CanAcceptPodForAgent(agentSlug string) bool {
	return r.CanAcceptPod() && r.SupportsAgent(agentSlug)
}
